File: 351003/Zhigimont/distcomp/publisher/src/main/java/com/bsuir/distcomp/service/WriterService.java
package com.bsuir.distcomp.service;

import com.bsuir.distcomp.dto.WriterRequestTo;
import com.bsuir.distcomp.dto.WriterResponseTo;
import com.bsuir.distcomp.entity.Writer;
import com.bsuir.distcomp.exception.EntityAlreadyExistsException;
import com.bsuir.distcomp.exception.EntityNotFoundException;
import com.bsuir.distcomp.mapper.WriterMapper;
import com.bsuir.distcomp.repository.WriterRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class WriterService implements CrudService<WriterRequestTo, WriterResponseTo>{

    private final WriterRepository repository;
    private final WriterMapper mapper;
    private final PasswordEncoder passwordEncoder;

    public WriterResponseTo create(WriterRequestTo dto) {
        Optional<Writer> writerOpt = repository.findByLogin(dto.getLogin());
        if(writerOpt.isPresent()){
            throw new EntityAlreadyExistsException("Writer with such login already exists");
        }
        Writer writer = mapper.toEntity(dto);
        writer.setPassword(passwordEncoder.encode(writer.getPassword()));
        Writer saved = repository.saveAndFlush(writer);
        return mapper.toDto(saved);
    }

    public List<WriterResponseTo> getAll() {
        return repository.findAll()
                .stream()
                .map(mapper::toDto)
                .toList();
    }

    public WriterResponseTo getById(Long id) {
        Writer writer = repository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Writer not found"));

        return mapper.toDto(writer);
    }

    public WriterResponseTo update(Long id, WriterRequestTo dto) {
        Writer existing = repository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Writer not found"));

        existing.setLogin(dto.getLogin());
        existing.setPassword(dto.getPassword());
        existing.setFirstname(dto.getFirstname());
        existing.setLastname(dto.getLastname());

        return mapper.toDto(repository.saveAndFlush(existing));
    }

    public void delete(Long id) {
        Writer writer = repository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Writer not found"));

        repository.delete(writer);
    }
}

File: 351003/Guzaev/publisher/app/services/writer_service.py
from repositories.base_repository import BaseRepository
from models.models import Writer
from dtos.writer_dto import WriterRequestTo, WriterResponseTo
from database import SessionLocal
from errors import AppError
from typing import List
from sqlalchemy.exc import IntegrityError

writer_repo = BaseRepository[Writer](Writer)

class WriterService:
    def create(self, dto: WriterRequestTo) -> WriterResponseTo:
        with SessionLocal() as db:
            entity = Writer(
                login=dto.login,
                password=dto.password,
                firstname=dto.firstname,
                lastname=dto.lastname
            )
            try:
                saved = writer_repo.create(db, entity)
            except IntegrityError:
                db.rollback()
                # Конфликт уникального логина → 403, как ждёт тест
                raise AppError(status_code=403, message="Login already exists", error_code=40301)

            return WriterResponseTo(
                id=saved.id,
                login=saved.login,
                firstname=saved.firstname,
                lastname=saved.lastname
            )


    def get_all(self) -> List[WriterResponseTo]:
        with SessionLocal() as db:
            entities = writer_repo.get_all(db)
            return [
                WriterResponseTo(
                    id=e.id, login=e.login,
                    firstname=e.firstname, lastname=e.lastname
                ) for e in entities
            ]

    def get_by_id(self, id: int) -> WriterResponseTo:
        with SessionLocal() as db:
            entity = writer_repo.get_by_id(db, id)
            if not entity:
                raise AppError(status_code=404, message="Writer not found", error_code=40401)
            return WriterResponseTo(
                id=entity.id, login=entity.login,
                firstname=entity.firstname, lastname=entity.lastname
            )

    def update(self, id: int, dto: WriterRequestTo) -> WriterResponseTo:
        with SessionLocal() as db:
            entity = writer_repo.get_by_id(db, id)
            if not entity:
                raise AppError(status_code=404, message="Writer not found", error_code=40402)

            existing = db.query(Writer).filter(Writer.login == dto.login, Writer.id != id).first()
            if existing:
                raise AppError(status_code=403, message="Login already exists", error_code=40302)

            # Обновляем поля сущности
            entity.login = dto.login
            entity.password = dto.password
            entity.firstname = dto.firstname
            entity.lastname = dto.lastname

            db.commit()
            db.refresh(entity)
            return WriterResponseTo(
                id=entity.id, login=entity.login,
                firstname=entity.firstname, lastname=entity.lastname
            )

    def delete(self, id: int) -> None:
        with SessionLocal() as db:
            success = writer_repo.delete(db, id)
            if not success:
                raise AppError(status_code=404, message="Writer not found", error_code=40403)
